presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -367,6 +376,24 @@ public void enforceMemoryLimits()
     }
 
     /**
+     * Enforce timeout at the query level
+     */
+    public void enforceQueryMaxRuntimeLimits()
+    {
+        Iterable<QueryExecution> runningQueries = queries.values().stream()
+                .filter(query -> query.getQueryInfo().getState() == RUNNING)
+                .collect(toImmutableList());
+
+        for (QueryExecution runningQuery : runningQueries) {
+            Duration queryMaxRuntime = SystemSessionProperties.getQueryMaxRuntime(runningQuery.getSession());
```

Comment:
Will it be confusing to fail a success query? And is there any overhead to fail a `Done` query?
I think @haozhun 's idea is more easier to understand.